src/iuse.cpp

```diff
@@ -2012,41 +2012,79 @@ int iuse::sew_advanced(player *p, item *it, bool, const tripoint& )
 
     // TODO: Wrap all the mods into structs, maybe even json-able
     // All possible mods here
-    std::array<std::string, 4> clothing_mods{
+    const std::array<const std::string, 4> clothing_mods{
         { "wooled", "furred", "leather_padded", "kevlar_padded" }
     };
 
     // Materials those mods use
-    std::array<std::string, 4> mod_materials{
+	const std::array<const std::string, 4> mod_materials{
         { "felt_patch", "fur", "leather", "kevlar_plate" }
     };
 
+    // TODO: Read data from \data\json\items\generic\filament.json
+    // Filament that can be used
+	const std::array<const std::string, 4> mod_filament{
+		{ "thread", "sinew", "plant_fibre", "yarn" }
+	};
+
+	const inventory &crafting_inv = p->crafting_inventory();
+
     // Cache available materials
-    std::map< itype_id, bool > has_enough;
+    std::map< itype_id, bool > has_enough_materials;
     const int items_needed = mod->volume() / 750_ml + 1;
-    const inventory &crafting_inv = p->crafting_inventory();
     // Go through all discovered repair items and see if we have any of them available
     for( auto &material : mod_materials ) {
-        has_enough[material] = crafting_inv.has_amount( material, items_needed );
+		has_enough_materials[material] = crafting_inv.has_amount( material, items_needed );
+    }
+
+	int filament_have_total = 0;
+
+	// Cache available filament
+	std::map< const itype_id, int > has_filament;
+	const int filament_needed = mod->volume() / 125_ml + 10; 	// We need extra filament to lose it on bad rolls
+    // Go through all filament items and write how much is available
+    for( auto &filament : mod_filament ) {
+
+		int filament_quantity = 0;
+
+		auto inv = p->crafting_inventory();
+		auto itm = inv.item_by_type(filament);
+
+		if (itm.count_by_charges())
+		{
+			filament_quantity = crafting_inv.charges_of(filament);
+		}
+		else
+		{
+			filament_quantity = crafting_inv.amount_of(filament);
```

Comment:
That was actually a debug code block which I wrote first to check available quantities (with 
```C++
p->add_msg_if_player(m_info, _("DEBUG: Caching filament=%s, quantity=%d, quantity_total=%d"), filament, filament_quantity, filament_have_total);
```
).

I will remove this block it as it is not really used.